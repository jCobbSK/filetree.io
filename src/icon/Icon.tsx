import React from 'react';
import { ReactComponent as FolderIcon } from './data/folder.svg';
import { ReactComponent as MenuIcon } from './data/menu.svg';
import { ReactComponent as MirrorIcon } from './data/mirror-glass.svg';

export enum IconType {
    FOLDER,
    MENU,
    MIRROR,
}

const ICON_MAP = {
    [IconType.FOLDER]: FolderIcon,
    [IconType.MENU]: MenuIcon,
    [IconType.MIRROR]: MirrorIcon,
};

interface Props {
    type: IconType;
    className?: string;
}

export const Icon: React.FC<Props> = ({ type, className }) => {
    const PickedIcon = ICON_MAP[type];
    return (
        <span className={className}>
            <PickedIcon />
        </span>
    );
};
