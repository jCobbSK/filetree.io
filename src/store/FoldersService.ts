import ASCIIFolder from '@pexxi/fold-to-ascii-ts';

export interface Folder {
    name: string;
    subFolders: Folder[];
}

const SEPARATOR = '/';
const HIGHLIGHTER = {
    START: '<mark>',
    END: '</mark>',
};

function createFolder(name: string): Folder {
    return {
        name,
        subFolders: [],
    };
}

function extractFirstFolderName(path: string): string[] {
    const [folderName, ...rest] = path.split(SEPARATOR);
    if (rest.length) {
        return [folderName, rest.join(SEPARATOR)];
    }
    return [folderName];
}

function getFolderByName(folders: Folder[], searchName?: string): Folder | undefined {
    if (searchName === undefined) {
        return undefined;
    }
    return folders.find(({ name }) => name === searchName);
}

function addPath(folders: Folder[], path?: string): Folder[] {
    if (path === undefined) {
        return folders;
    }
    const [folderName, nextPath] = extractFirstFolderName(path);
    const folder = getFolderByName(folders, folderName);
    if (folder) {
        addPath(folder.subFolders, nextPath);
        return folders;
    }
    const newFolder = createFolder(folderName);
    folders.push(newFolder);
    addPath(newFolder.subFolders, nextPath);
    return folders;
}

function doesFitQuery(folderName: string, query: string): boolean {
    return ASCIIFolder.foldMaintaining(folderName).toLowerCase().includes(query);
}

const folderMatchMap = (query: string) => (folder: Folder): Folder | undefined => {
    if (doesFitQuery(folder.name, query)) {
        return folder;
    }

    const subFolders = folder.subFolders
        .map(folderMatchMap(query))
        .filter((folder) => !!folder) as Folder[];
    if (!subFolders.length) {
        return undefined;
    }
    return {
        ...folder,
        subFolders,
    };
};

function getQueryOccuranceIndexes(folderName: string, query: string): number[] {
    const foldedFolderName = ASCIIFolder.foldMaintaining(folderName).toLowerCase();
    const result = [];
    let index = foldedFolderName.indexOf(query);
    while (index !== -1) {
        result.push(index);
        index = foldedFolderName.indexOf(query, index + 1);
    }
    return result;
}

export function getParsedFoldersFromPaths(paths: string[]): Folder[] {
    return paths.reduce<Folder[]>(addPath, []);
}

export function getFilteredFolders(folders: Folder[], query: string): Folder[] {
    return folders.map(folderMatchMap(query)).filter((folder) => !!folder) as Folder[];
}

export function getMarkedFolderNameWithQuery(folderName: string, query: string): string {
    if (!query) {
        return folderName;
    }
    const queryStartIndexes = getQueryOccuranceIndexes(folderName, query);
    const queryLength = query.length;
    const queryEndIndexes = queryStartIndexes.map((index) => index + queryLength);
    if (!queryStartIndexes.length) {
        return folderName;
    }

    return [...folderName.split(''), ''].reduce<string>(
        (result, folderCharacter, folderNameIndex) => {
            if (queryStartIndexes.includes(folderNameIndex)) {
                return `${result}${HIGHLIGHTER.START}${folderCharacter}`;
            }
            if (queryEndIndexes.includes(folderNameIndex)) {
                return `${result}${HIGHLIGHTER.END}${folderCharacter}`;
            }
            return `${result}${folderCharacter}`;
        },
        '',
    );
}
