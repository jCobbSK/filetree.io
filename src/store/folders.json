{
    "paths": [
        "Adidas",
        "Bata",
        "CCC",
        "Nike/T-shirts/Shirt Green",
        "Nike/Boots",
        "Nike/Nike Jordans Black",
        "Nike/T-shirts/Shirt Black",
        "Nike/T-shirts/Shirt White",
        "Zara/FancyJacket",
        "Zara/Leather Shirt"
    ]
}
