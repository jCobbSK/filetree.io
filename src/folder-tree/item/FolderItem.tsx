import React from 'react';

import { getMarkedFolderNameWithQuery } from '../../store/FoldersService';
import { Icon, IconType } from '../../icon/Icon';
import { Menu } from './menu/Menu';
import style from './FolderItem.module.css';
import { RenameInput } from './RenameInput';
import { useFolderItem } from './useFolderItem';

interface Props {
    id: string;
    name: string;
    query: string;
}

export const FolderItem: React.FC<Props> = ({ id, name, query }) => {
    const {
        isRenaming,
        isMenuVisible,
        isMenuIconVisible,
        onMenuIconClick,
        onMenuItemClick,
        onRenameMenuItemClick,
        onRename,
        onHover,
        onHoverOut,
    } = useFolderItem(id);

    return (
        <div className={style.wrapper} onMouseEnter={onHover} onMouseLeave={onHoverOut}>
            <Icon type={IconType.FOLDER} className={style.icon} data-cy="folder-icon" />
            {!isRenaming && (
                <span
                    data-cy="folder-name"
                    dangerouslySetInnerHTML={{ __html: getMarkedFolderNameWithQuery(name, query) }}
                />
            )}
            {isRenaming && <RenameInput name={name} onConfirm={onRename} />}
            {isMenuIconVisible && (
                <button onClick={onMenuIconClick} className={style.menuIcon} data-cy="menu-button">
                    <Icon type={IconType.MENU} />
                </button>
            )}
            {isMenuVisible && (
                <Menu id={id} onClick={onMenuItemClick} onRenameClick={onRenameMenuItemClick} />
            )}
        </div>
    );
};
