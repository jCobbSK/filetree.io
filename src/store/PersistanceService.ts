import { Folder } from './FoldersService';

const LOCAL_STORAGE_KEY = 'folders';

export function loadFolders(): Folder[] | undefined {
    const stringifiedFolders = localStorage.getItem(LOCAL_STORAGE_KEY);
    if (stringifiedFolders) {
        return JSON.parse(stringifiedFolders);
    }
    return undefined;
}
export function saveFolders(folders: Folder[]): void {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(folders));
}
