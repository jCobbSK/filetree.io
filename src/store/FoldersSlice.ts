import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { paths } from './folders.json';
import {
    getParsedFoldersFromPaths,
    getFilteredFolders,
    Folder,
    findFolderById,
    createFolder,
    getFoldersWithoutFolderWithId,
} from './FoldersService';
import { StoreType } from './store';

interface State {
    folders: Folder[];
}

const initialState: State = {
    folders: getParsedFoldersFromPaths(paths),
};

const foldersSlice = createSlice({
    name: 'folders',
    initialState,
    reducers: {
        addChild(state: State, { payload }: PayloadAction<string>) {
            const folder = findFolderById(state.folders, payload);
            if (folder) {
                folder.subFolders.push(createFolder('Newly added folder'));
            }
        },
        removeFolder(state: State, { payload }: PayloadAction<string>) {
            state.folders = getFoldersWithoutFolderWithId(state.folders, payload);
        },
        renameFolder(state: State, { payload }: PayloadAction<{ id: string; newName: string }>) {
            const folder = findFolderById(state.folders, payload.id);
            if (folder) {
                folder.name = payload.newName;
            }
        },
    },
});

export const {
    reducer: foldersReducer,
    actions: { addChild, removeFolder, renameFolder },
} = foldersSlice;

export const selectFilteredFolders = (query: string) => (store: StoreType) => {
    if (!query) {
        return store.foldersReducer.folders;
    }
    return getFilteredFolders(store.foldersReducer.folders, query);
};
