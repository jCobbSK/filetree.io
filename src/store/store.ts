import { configureStore } from '@reduxjs/toolkit';

import { foldersReducer } from './FoldersSlice';

export const store = configureStore({
    reducer: {
        foldersReducer: foldersReducer,
    },
});
export type StoreType = ReturnType<typeof store.getState>;
