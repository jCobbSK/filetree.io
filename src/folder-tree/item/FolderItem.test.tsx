import React from 'react';
import { shallow } from 'enzyme';

import * as FolderTreeService from '../../store/FoldersService';
import { IconType } from '../../icon/Icon';
import { FolderItem } from './FolderItem';

describe('<FolderItem />', () => {
    it('renders icon and highlighted name with query', () => {
        const getMarkedFolderNameSpy = jest
            .spyOn(FolderTreeService, 'getMarkedFolderNameWithQuery')
            .mockReturnValueOnce('NikeTransformed');

        const wrapper = shallow(<FolderItem id="uuid" name="Nike" query="N" />);

        expect(getMarkedFolderNameSpy).toHaveBeenCalledWith('Nike', 'N');
        expect(wrapper.find('[data-cy="folder-icon"]')).toHaveProp({
            type: IconType.FOLDER,
            className: 'icon',
        });
        expect(wrapper.find('[data-cy="folder-name"]')).toHaveProp({
            dangerouslySetInnerHTML: {
                __html: 'NikeTransformed',
            },
        });
    });

    it('renders menu icon when mouse is entered over component', () => {
        const wrapper = shallow(<FolderItem id="uuid" name="Nike" query="N" />);

        wrapper.simulate('mouseEnter');

        expect(wrapper.find('[data-cy="menu-button"]')).toHaveLength(1);
        expect(wrapper.find('[data-cy="menu-button"] > Icon')).toHaveLength(1);
    });

    it('renders menu when menu icon is clicked', () => {
        const wrapper = shallow(<FolderItem id="uuid" name="Nike" query="N" />);
        wrapper.simulate('mouseEnter');

        wrapper.find('[data-cy="menu-button"]').simulate('click');

        expect(wrapper.find('Menu')).toHaveProp({
            id: 'uuid',
            name: 'Nike',
        });
    });
});
