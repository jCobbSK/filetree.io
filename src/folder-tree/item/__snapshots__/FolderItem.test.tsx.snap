// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FolderItem /> renders menu icon when isMenuIconVisible is true 1`] = `
<div
  className="wrapper"
  onMouseEnter={[MockFunction onHover]}
  onMouseLeave={[MockFunction onHoverOut]}
>
  <Icon
    className="icon"
    data-cy="folder-icon"
    type={0}
  />
  <span
    dangerouslySetInnerHTML={
      Object {
        "__html": "Nike",
      }
    }
    data-cy="folder-name"
  />
  <button
    className="menuIcon"
    data-cy="menu-button"
    onClick={[MockFunction onMenuIconClick]}
  >
    <Icon
      type={1}
    />
  </button>
</div>
`;

exports[`<FolderItem /> renders menu when isMenuVisible is true 1`] = `
<div
  className="wrapper"
  onMouseEnter={[MockFunction onHover]}
  onMouseLeave={[MockFunction onHoverOut]}
>
  <Icon
    className="icon"
    data-cy="folder-icon"
    type={0}
  />
  <span
    dangerouslySetInnerHTML={
      Object {
        "__html": "Nike",
      }
    }
    data-cy="folder-name"
  />
  <Menu
    id="uuid"
    onClick={[MockFunction onMenuItemClick]}
    onRenameClick={[MockFunction onRenameMenuItemClick]}
  />
</div>
`;

exports[`<FolderItem /> renders rename input when isRenaming is true 1`] = `
<div
  className="wrapper"
  onMouseEnter={[MockFunction onHover]}
  onMouseLeave={[MockFunction onHoverOut]}
>
  <Icon
    className="icon"
    data-cy="folder-icon"
    type={0}
  />
  <RenameInput
    name="Nike"
    onConfirm={[MockFunction onRename]}
  />
</div>
`;

exports[`<FolderItem /> renders wrapper with hover listeners, icon and highlighted name with query when isRenaming is false 1`] = `
<div
  className="wrapper"
  onMouseEnter={[MockFunction onHover]}
  onMouseLeave={[MockFunction onHoverOut]}
>
  <Icon
    className="icon"
    data-cy="folder-icon"
    type={0}
  />
  <span
    dangerouslySetInnerHTML={
      Object {
        "__html": "NikeTransformed",
      }
    }
    data-cy="folder-name"
  />
</div>
`;
